Jg/polished 3.1 vcf export

jkgoodrich opened the pull request:
This PR is to add the cleaned up gnomad v3.1 VCF export code incorporating in the VCF export of the HGDP + TGP subset. It does rely on the sanity code being in place, but that is a separate PR and I was hoping to have this PR looked at in the meantime because of how long it is. Therefore the checks will currently fail until we get the sanity check PR in master and another PR for the generalized functions in gnomad_methods.

The two tests I performed used the sanity check files from the following branches:
https://github.com/broadinstitute/gnomad_methods/tree/jg/temp_sanity_check
https://github.com/broadinstitute/gnomad_qc/tree/jg/temp_sanity_check

It also relies on the code added to gnomad_methods in the following PR:
https://github.com/broadinstitute/gnomad_methods/pull/365

Output from the tests of the code here: 

Full dataset: 
https://console.cloud.google.com/dataproc/jobs/2be4d656f7824125906e3ac7d4dff6c0?region=us-central1&project=broad-mpg-gnomad

Hgdp + tgp subset: https://console.cloud.google.com/dataproc/jobs/c62c14d696ce4e0c956cf09482a2107a?region=us-central1&project=broad-mpg-gnomad

ch-kr commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+
+            logger.info("Running check on VCF fields and info dict...")
+            if not vcf_field_check(prepared_vcf_ht, header_dict, new_row_annots):
+                logger.error("Did not pass VCF field check")
```
would you not want to exit the script here (rather than just logger)?

ch-kr commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+                t = filter_to_test(t)
+
+            if chromosome:
+                t = hl.filter_intervals(t, [hl.parse_locus_interval(chromosome)])
```
do you need to specify that the reference is `export_reference` here?

jkgoodrich replied:
This is done before `rekey_new_reference(t, export_reference` though so I think this is OK

ch-kr commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    )
+    parser.add_argument(
+        "--test",
+        help="Create release files using only 5 partitions on chr20, chrX, and chrY for testing purposes",
```
```suggestion
        help="Create release files using only 2 partitions on chr20, chrX, and chrY for testing purposes",
```
the default for `filter_to_test` is 2 partitions

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    remove_fields_from_globals,
+)
+
+hl.stop()
```
I think you actually already fixed this, it isn't needed anymore I need to remove it

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    "transmitted_singleton",
+]
+SITE_FIELDS.extend(NEW_SITE_FIELDS)
+AS_FIELDS.append("AS_SB_TABLE")
```
Yes, I have a few updates to make after additions that Grace recently added to gnomad_methods https://github.com/broadinstitute/gnomad_methods/commit/5a02fc70b1003f77b594029dee43c1cd58b986da#diff-b5a0be260e1a0153ac1535bd57a840df4cef8fadb1b0349fabbced3881d1f85f

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+AS_FIELDS.append("AS_SB_TABLE")
+
+# Remove original alleles for containing non-releasable alleles
+MISSING_ALLELE_TYPE_FIELDS = ["original_alleles", "has_star"]
```
no, `has_star` is missing for variants in gnomAD v3.1 (and v3) so we remove it unlike in UKBB

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+
+# Make subset list (used in properly filling out VCF header descriptions and naming VCF info fields)
+SUBSET_LIST_FOR_VCF = SUBSETS.copy()
+SUBSET_LIST_FOR_VCF.append("")
```
yes

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+        "RGQ": {
+            "Number": "1",
+            "Type": "Integer",
+            "Description": "Unconditional reference genotype confidence, encoded as a phred quality -10*log10 p(genotype call is wrong)",
```
Not sure, I took from here: https://gatk.broadinstitute.org/hc/en-us/articles/360035531692-VCF-Variant-Call-Format and that is how they have it. I prefer to defer to the GATK descriptions

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    # Add AS annotations to info dict
+    for field in AS_FIELDS:
+        vcf_info_dict[field] = t["release_ht_info"][f"{field}"]
+    for field in VQSR_FIELDS + AS_VQSR_FIELDS:
```
oops, yes I forgot I changed it, thank you!

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+            }
+            expr_dict.update(combo_dict)
+
+    if not is_subset:
```
We drop age info for the subset so this hist is dropped

jkgoodrich commented on gnomad_qc/v3/resources/release.py:
```diff
@@ -44,3 +55,95 @@ def release_ht_path(
         return f"gs://gnomad/release/{release_version}/ht/gnomad.{data_type}.r{release_version}.sites.ht"
+
+
+def release_var_hist_path(data_source: str, freeze: int) -> str:
```
oops, nope, this shouldn't be here and is not used

jkgoodrich commented on gnomad_qc/v3/resources/constants.py:
```diff
@@ -1,5 +1,7 @@
-CURRENT_RELEASE = "3.1"
+RELEASES = ["3", "3.1", "3.1.1"]
+CURRENT_RELEASE = "3.1.1"
+HGDP_TGP_RELEASES = ["3.1"]
```
We don't have an updated v3.1.1 HGDP + TGP subset yet, I will update that when thy exist

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+            ),  # TODO: Change to release_sites().ht(),
+        }
+        # Downsampling and subset entries to remove from VCF's freq export
+        # Note: Need to extract the non-standard downsamplings from the freq_meta struct to the FREQ_ENTRIES_TO_REMOVE
```
If a population has less samples than one of the down sampling numbers it will get downsampled to it's max sample size instead of one of the standard downsamplings in the `DOWNSAMPLINGS` global

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    new_vcf_info_dict = {}
+    for i, j in vcf_info_dict.items():
+        i = i.replace("_adj", "")
+        i = i.replace("-", "_")  # VCF 4.3 specs do not allow hyphens in info fields
```
Some of the subsets have a "_" in their names so we needed in order to not split on part of the subset name we switched to a "-" delimiter

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    filter_dict = make_vcf_filter_dict(
+        hl.eval(t.filtering_model.snv_cutoff.min_score),
+        hl.eval(t.filtering_model.indel_cutoff.min_score),
+        inbreeding_cutoff=INBREEDING_COEFF_HARD_CUTOFF,
```
will change to take in an argument incase we generalize this function

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+        )
+
+    # Select relevant fields for VCF export
+    ht = ht.select("info", "filters", "rsid")
```
Yes, we do not have a qual annotation. I think for UKBB we ended up using qualapprox right? I can add a note, but I would like to keep as is since v3 and v3.1 didn't have this

mike-w-wilson commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    hl.init(
+        log="/vcf_release.log",
+        default_reference="GRCh38",
+        tmp_dir="hdfs:///vcf_write.tmp/",
```
This may be updated by now, I've just always added it out of habit. Historically you needed it to use the SSDs but that may not be the case anymore: https://atgu.slack.com/archives/CNE92QZMF/p1542132816013200

ch-kr commented on gnomad_qc/v3/utils.py:
```diff
@@ -22,19 +22,22 @@ def remove_fields_from_globals(global_field: List[str], fields_to_remove: List[s
     """
     ref = hl.get_reference("GRCh38")
-    my_contigs = [f"chr{i}" for i in range(1, 23)] + ["chrX", "chrY", "chrM"]
+    standard_contigs = [f"chr{i}" for i in range(1, 23)] + ["chrX", "chrY", "chrM"]
```
it just occurred to me, maybe we should add an option here to specify which contigs to release (in case chrM shouldn't always be included here, for example)

ch-kr commented on gnomad_qc/v3/utils.py:
```diff
@@ -22,19 +22,22 @@ def remove_fields_from_globals(global_field: List[str], fields_to_remove: List[s
             logger.info(f"'{field}'' missing from {global_field}")
 
 
+# TODO: maybe list one of the non-standard contigs?
```
here's an example from a ukbb vcf: `##contig=<ID=chr3_GL000221v1_random,length=155397,assembly=GRCh38>`

ch-kr commented on gnomad_qc/v3/utils.py:
```diff
@@ -70,3 +73,28 @@ def rekey_new_reference(
+
+# TODO: Needs to be modified to fix the problem of the alt het correction on a split MT
+def hom_alt_depletion_fix(
+        mt: hl.MatrixTable,
```
these next few lines seem to be indented slightly too much?

ch-kr commented on gnomad_qc/v3/utils.py:
```diff
@@ -70,3 +73,28 @@ def rekey_new_reference(
+) -> hl.MatrixTable:
+    """
+    Adjust MT genotypes with temporary fix for the depletion of homozygous alternate genotypes.
+
```
might be helpful to add this link to the docstring: https://gnomad.broadinstitute.org/blog/2020-10-gnomad-v3-1-new-content-methods-annotations-and-data-availability/#tweaks-and-updates

ch-kr commented on gnomad_qc/v3/utils.py:
```diff
@@ -70,3 +73,28 @@ def rekey_new_reference(
+    """
+    Adjust MT genotypes with temporary fix for the depletion of homozygous alternate genotypes.
+
+    :param mt: Input MT to add hom alt depletion genotype fix to
```
```suggestion
    :param mt: Input MT that needs hom alt genotype fix
```

ch-kr commented on gnomad_qc/v3/utils.py:
```diff
@@ -70,3 +73,28 @@ def rekey_new_reference(
+    Adjust MT genotypes with temporary fix for the depletion of homozygous alternate genotypes.
+
+    :param mt: Input MT to add hom alt depletion genotype fix to
+    :param af_expr: Allele frequency expression to determine variants that need the hom alt fix
```
```suggestion
    :param af_expr: Allele frequency expression to determine which variants need the hom alt fix
```

ch-kr commented on gnomad_qc/v3/utils.py:
```diff
@@ -70,3 +73,28 @@ def rekey_new_reference(
+    """
+    return mt.annotate_entries(
+        GT=hl.cond(
+            (af_expr > af_cutoff) & mt.GT.is_het() & ((mt.AD[1] / mt.DP) > ab_cutoff),
```
Tim suggested putting the `mt.GT.is_het()` as the first filter here for efficiency

ch-kr replied:
the only other condition I added for the fix here was this `& ~mt.het_non_ref` (from adding this before splitting`mt = mt.annotate_entries(het_non_ref=mt.LGT.is_het_non_ref())`)

ch-kr commented on gnomad_qc/v3/resources/release.py:
```diff
@@ -59,15 +62,22 @@ def release_ht_path(
     """
-    return f"{get_release_path(data_source, freeze)}/json/{data_source}.freeze_{freeze}.json"
+    return VersionedTableResource(
+        CURRENT_RELEASE,
```
adding`default_version=CURRENT_RELEASE` might make this clearer

ch-kr commented on gnomad_qc/v3/resources/release.py:
```diff
@@ -59,15 +62,22 @@ def release_ht_path(
-    return f"{get_release_path(data_source, freeze)}/json/{data_source}.freeze_{freeze}.json"
+    return VersionedTableResource(
+        CURRENT_RELEASE,
+        {
```
same comment here (adding `versions=` might be clearer)

ch-kr commented on gnomad_qc/v3/resources/release.py:
```diff
@@ -59,15 +62,22 @@ def release_ht_path(
+            release: TableResource(
+                path=release_ht_path(release_version=release, public=public)
+            )
+            for release in RELEASES
```
this isn't imported yet, right?

ch-kr reviewed (commented):
a few more comments -- sorry I started reviewing already because I thought you re-requested review (I got the two vcf PRs mixed in my head)

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    for hist in HISTS:
+        for hist_type in [f"{hist_prefix}qual_hists", f"{hist_prefix}raw_qual_hists"]:
+            hist_name = hist
+            if "raw" in hist_type:
```
True, I have removed, let me know if it looks OK

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    )
+
+    # Add variant quality histograms to info dict
+    vcf_info_dict.update(make_hist_dict(bin_edges, adj=True))
```
Mike looked into this: no previous gnomAD release did, HT or VCF, but it looks like we added them to the v3.1 release HT, https://the-tgg.slack.com/archives/CNLMF0JLV/p1603402420087000 and didn't grab them on VCF export. I think this is likely fine not to have in the VCF as we have not typically supplied them and they are on the HT if users would like to use them

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    logger.info("Unfurling freq data...")
+
+    if is_subset and gnomad_release:
+        prefix = f"gnomad-"
```
yes, like all of our other globals these globals also have an '_'. The frequencies have a `-` because of our subsets having an `_`

jkgoodrich reviewed (commented):
I'm still working on this @ch-kr so not ready for a full review again, but wanted to add some of my comments now

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    Clean up the Table returned by `prepared_vcf_ht` so it is ready for export to a VCF with `hl.export_vcf`.
+
+    Specifically:
+        - Drop histograms and frequency entries that are not wanted in the VCF.
```
Sounds good, I will add a note for v4

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,1019 @@
+    if not is_subset or (is_subset and gnomad_release):
+        combo_dict = {
+            f"{prefix}popmax": t[popmax].pop,
+            f"{prefix}AC_popmax": t[popmax].AC,
```
Because of what is returned by `make_info_dict` when popmax is True: https://github.com/broadinstitute/gnomad_methods/blob/e6f8e5f03cab4e5c36af983e1d81208f2c208ac0/gnomad/utils/vcf.py#L568 However, I could change the common code to use `label_delimiter` instead of `_` in the common code if that would be better

jkgoodrich commented on gnomad_qc/v3/resources/basics.py:
```diff
@@ -67,6 +67,9 @@ def get_gnomad_v3_mt(
             "gs://gnomad/raw/genomes/3.1/gnomad_v3.1_sparse_unsplit.repartitioned.mt"
         ),
+        "3.1.1": MatrixTableResource(
+            "gs://gnomad/raw/genomes/3.1/gnomad_v3.1_sparse_unsplit.repartitioned.mt"
```
We need both so that the resource will work, but I modified it a little to use 3.1 for 3.1.1, let me know if that is better or not

ch-kr commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -410,32 +371,59 @@ def make_info_expr(
+
+       One and only one of `gnomad_full_release`, `subset_release`, or `gnomad_full_for_subset` must be True.
+
+    If `gnomad_full_release` is True the following will be unfurled:
```
this is so much clearer, thank you!!

ch-kr commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,974 @@
+            ),
+            "drop_hists": None,
+            "include_age_hists": False,
+            "subset_pops": {"gnomad": POPS, "": HGDP_TGP_POPS},
```
maybe we can update the inputs to `populate_info_dict`? `subset_pops` defaults to `POPS` there but maybe it should default to `HGDP_TGP_POPS`?

jkgoodrich replied:
It's really only used in sanity checks, populate_info_dict doesn't use it

ch-kr replied:
`populate_info_dict` uses subset_pops on line 271, right? https://github.com/broadinstitute/gnomad_qc/blob/jg/polished_3.1_vcf_export/gnomad_qc/v3/create_release/prepare_vcf_data_release.py#L274
it sets `pops = subset_pops` if subset isn't "gnomAD"

jkgoodrich replied:
Yes, but no, we renamed that argument to subset_pops I think, but it doesn't use these. Honestly, at this point I need to look over it all again including actually running the code through the test to make sure things look as expected so I think I will do the other comments and then try running the tests to see what it looks like.

jkgoodrich replied:
Renamed this to `sample_sum_sets_and_pops` because that is the parameter used in the validity checks and this is only used there

ch-kr reviewed (changes requested):
a few more small things. I also feel we can simplify the pops vs subset pops a little more and will keep thinking about it

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,974 @@
+    new_row_annots = []
+    for x in row_annots:
+        x = x.replace("-adj", "")
+        x = x.replace("-", "_")  # VCF 4.3 specs do not allow hyphens in info fields
```
Yeah, we changed because we got reports from users that they were having trouble with the `-` in some programs they were trying to use so we figured we might as well put it in

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,974 @@
+    # Reformat AS_SB_TABLE for use in ht_to_vcf_mt
+    ht = ht.annotate(
+        info=ht.info.annotate(
+            AS_SB_TABLE=hl.array([ht.info.AS_SB_TABLE[:2], ht.info.AS_SB_TABLE[2:]])
```
I will add a TODO, but for sake of not regenerating the v3.1 info HT I will keep this as is

jkgoodrich commented on gnomad_qc/v3/utils.py:
```diff
@@ -1,25 +1,37 @@
     return mt.annotate_entries(
         GT=hl.cond(
-            (af_expr > af_cutoff) & mt.GT.is_het() & ((mt.AD[1] / mt.DP) > ab_cutoff),
+            (mt.GT.is_het() & ~het_non_ref_expr & af_expr > af_cutoff)
```
I'm very confused because I absolutely remember changing this to use your exact code

jkgoodrich commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -0,0 +1,974 @@
+                logger.info(
+                    "Loading the HGDP + TGP subset MT and annotating with the prepared VCF HT for VCF export..."
+                )
+                t = hgdp_1kg_subset(dense=True).mt().select_rows()
```
I read in the HT not the MT

ch-kr commented on gnomad_qc/v3/resources/release.py:
```diff
@@ -91,8 +177,8 @@ def hgdp_1kg_subset_annotations(sample: bool = True) -> VersionedTableResource:
     return VersionedTableResource(
-        CURRENT_RELEASE,
-        {
+        default_version=CURRENT_RELEASE,
```
sorry, starting to lose track of what I've already asked. should this be `CURRENT_HGDP_TGP_RELEASE`?

ch-kr replied:
I also just noticed there's a `3.1` and `v3.1` here 
![image](https://user-images.githubusercontent.com/19142091/121722153-11b58e80-cab3-11eb-96ee-ca88fd21b388.png)

jkgoodrich replied:
yes I think when Grace made the release the first time she accidentally added the `v` I have this listed on my storage sheet asking if we can just delete

ch-kr commented on gnomad_qc/v3/resources/release.py:
```diff
@@ -63,6 +63,92 @@ def release_ht_path(
+        sites VCF path
+    :return: Filepath for the desired VCF
+    """
+    if hgdp_tgp_subset:
```
should this function also have this check? `if release_version not in HGDP_TGP_RELEASES:`

ch-kr reviewed (commented):
a few more small things (a couple questions about the release resource file and same question about subset_pops)

jkgoodrich commented:
I have addressed all comments, but I want to run a full test on the code again. I think it makes sense to wait until the sanity check code is updated and merged so that I can run it with the new sanity check code.

jkgoodrich commented:
After a few fixes that came up during testing, this is back to you @ch-kr.

Output from the tests of the code here:

Full callset test: https://console.cloud.google.com/dataproc/jobs/d96ead586837410ca1c036756e2216f1?region=us-central1&project=broad-mpg-gnomad
HGDP + TGP subset test: https://console.cloud.google.com/dataproc/jobs/308f2e02543d49b5bd47cad970979a0d?region=us-central1&project=broad-mpg-gnomad

ch-kr commented on gnomad_qc/v3/resources/release.py:
```diff
@@ -58,9 +58,9 @@ def release_ht_path(
     """
     if public:
-        return f"gs://gnomad-public/release/{release_version}/ht/{data_type}/gnomad.{data_type}.r{release_version}.sites.ht"
+        return f"gs://gnomad-public/release/{release_version}/ht/{data_type}/gnomad.{data_type}.v{release_version}.sites.ht"
```
```suggestion
        return f"gs://gnomad-public-requester-pays/release/{release_version}/ht/{data_type}/gnomad.{data_type}.v{release_version}.sites.ht"
```
(though this won't work for v3.0)

jkgoodrich replied:
All resources are currently not working as expected. There is a good chance things will be moving around more so I am waiting for a full and separate PR to address all of these issues.

ch-kr commented on gnomad_qc/v3/resources/release.py:
```diff
@@ -58,9 +58,9 @@ def release_ht_path(
+        return f"gs://gnomad-public/release/{release_version}/ht/{data_type}/gnomad.{data_type}.v{release_version}.sites.ht"
     else:
-        return f"gs://gnomad/release/{release_version}/ht/gnomad.{data_type}.r{release_version}.sites.ht"
+        return f"gs://gnomad/release/{release_version}/ht/{data_type}/gnomad.{data_type}.v{release_version}.sites.ht"
```
I think this currently doesn't work for any of the 3 releases (3.1 and 3.1.1 don't have hts in this bucket, and 3.0 has an 'r' instead of a 'v') -- not sure if this matters right now and will be fixed later

ch-kr commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -360,6 +364,7 @@ def make_info_expr(
 
     vcf_info_dict["revel_score"] = t["revel"]["revel_score"]
 
+    # In the v3.1 in silico files this was max_ds, but changed to splice_ai_score in releases after v3.1
```
```suggestion
    # In the v3.1 release in silico files this was max_ds, but changed to splice_ai_score in releases after v3.1
```

ch-kr commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -168,7 +172,7 @@ def populate_subset_info_dict(
     :param description_text: Text describing the sample subset that should be added to the INFO description.
     :param pops: Dict of sample global population names for gnomAD genomes. Default is POPS.
-    :param faf_pops: List of faf population names. Default is FAF_POPS.
+    :param faf_pops: Dict of faf population names. Default is FAF_POPS.
```
```suggestion
    :param faf_pops: Dict with faf pop names (keys) and descriptions (values).  Default is FAF_POPS.
```

ch-kr commented on gnomad_qc/v3/create_release/prepare_vcf_data_release.py:
```diff
@@ -248,7 +252,7 @@ def populate_info_dict(
         is 'gnomad', in that case `gnomad_pops` is used. Default is POPS.
     :param gnomad_pops: Dict of sample global population names for gnomAD genomes. Default is POPS.
-    :param faf_pops: List of faf population names. Default is FAF_POPS.
+    :param faf_pops: Dict of faf population names. Default is FAF_POPS.
```
```suggestion
    :param faf_pops: Dict with faf pop names (keys) and descriptions (values).  Default is FAF_POPS.
```

ch-kr reviewed (commented):
a few small suggestions, but this looks pretty good. I have a couple questions about the validity checks output, though:

first, a few pops seem to be missing (in both the full callset and hgdp/tgp subset test outputs):
```
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-papuan-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-biakapygmy-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-san-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-mbutipygmy-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-melanesian-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-papuan-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-biakapygmy-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-san-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-mbutipygmy-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-melanesian-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-papuan-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-biakapygmy-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-san-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-mbutipygmy-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-melanesian-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-papuan-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-papuan-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-biakapygmy-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-biakapygmy-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-san-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-san-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-mbutipygmy-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-mbutipygmy-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-melanesian-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AC-hgdp-melanesian-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-papuan-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-papuan-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-biakapygmy-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-biakapygmy-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-san-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-san-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-mbutipygmy-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-mbutipygmy-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-melanesian-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): AN-hgdp-melanesian-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-papuan-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-papuan-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-biakapygmy-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-biakapygmy-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-san-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-san-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-mbutipygmy-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-mbutipygmy-XY-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-melanesian-XX-adj is not in table's info field
WARNING (gnomad.assessment.validity_checks 177): nhomalt-hgdp-melanesian-XY-adj is not in table's info field
``` 
are some populations not being exported correctly? is this something that was missed just when creating the vcf ht/header? these pop fields don't get flagged in the info field checks (which check the info fields against the pickled header):
```
INFO (vcf_release 914): Running check on VCF fields and info dict...
WARNING (gnomad.assessment.validity_checks 826): SB in info field does not exist in VCF header!
INFO (gnomad.assessment.validity_checks 842): Passed VCF fields check!
```

second, do you know why some AFs are being flagged for missingness checks when the AC and ANs are passing? for example:
![image](https://user-images.githubusercontent.com/19142091/129102268-9aaa7d2d-75b8-43f0-96b2-a4c20bd09560.png)

jkgoodrich commented:
For the first comment, this is because of gnomAD QC fully removing some subpopulations during the QC outlier detection. It is a known issue that we hope to improve in future QC with highly diverse samples. So all of these are currently expected.

I will look into the second comment

jkgoodrich commented:
So my guess for the missingness is that if there is an AN = 0 then the AF evaluates to NA because of the divide by zero. I'm OK keeping as is because it really is a divide by zero, but would like to hear other thoughts @gtiao? For this case we could also change it so it is a struct with all missing or a struct with all zero if that seems easier to interpret? Also, I think this is something that needs to be handled in the frequency code or the release HT/MT code so would be better fixed in a different PR and I can add a separate issue.

ch-kr commented:
that makes sense to me -- just approved
